llms.txt

```diff
@@ -11289,6 +11290,780 @@ Doc-Content: https://docs.kluster.ai/tutorials/klusterai-api/image-analysis.ipyn
+      "            \"function\": {\n",
+      "              \"arguments\": \"{\\\"expression\\\": \\\"1337 * 42\\\"}\",\n",
+      "              \"name\": \"calculator\"\n",
+      "            },\n",
+      "            \"type\": \"function\"\n",
+      "          }\n",
+      "        ]\n",
+      "      },\n",
+      "      \"stop_reason\": 128008\n",
+      "    }\n",
+      "  ],\n",
+      "  \"created\": 1746741970,\n",
+      "  \"model\": \"klusterai/Meta-Llama-3.1-8B-Instruct-Turbo\",\n",
+      "  \"object\": \"chat.completion\",\n",
+      "  \"service_tier\": null,\n",
+      "  \"system_fingerprint\": null,\n",
+      "  \"usage\": {\n",
+      "    \"completion_tokens\": 20,\n",
+      "    \"prompt_tokens\": 252,\n",
+      "    \"total_tokens\": 272,\n",
+      "    \"completion_tokens_details\": null,\n",
+      "    \"prompt_tokens_details\": null\n",
+      "  },\n",
+      "  \"prompt_logprobs\": null\n",
+      "}\n"
+     ]
+    }
+   ],
+   "source": [
+    "def run_with_tools(prompt, tools, model=\"klusterai/Meta-Llama-3.1-8B-Instruct-Turbo\"):\n",
+    "    messages = [\n",
+    "        {\"role\": \"user\", \"content\": prompt}\n",
+    "    ]\n",
+    "    \n",
+    "    response = client.chat.completions.create(\n",
+    "        model=model,\n",
+    "        messages=messages,\n",
+    "        tools=tools,\n",
+    "        tool_choice=\"auto\"\n",
+    "    )\n",
+    "    \n",
+    "    return response\n",
+    "\n",
+    "# Define a calculator tool\n",
+    "calculator_tools = [\n",
+    "    {\n",
+    "        \"type\": \"function\",\n",
+    "        \"function\": {\n",
+    "            \"name\": \"calculator\",\n",
+    "            \"description\": \"Perform arithmetic calculations\",\n",
+    "            \"parameters\": {\n",
+    "                \"type\": \"object\",\n",
+    "                \"properties\": {\n",
+    "                    \"expression\": {\n",
+    "                        \"type\": \"string\",\n",
+    "                        \"description\": \"The mathematical expression to evaluate\"\n",
+    "                    }\n",
+    "                },\n",
+    "                \"required\": [\"expression\"]\n",
+    "            }\n",
+    "        }\n",
+    "    }\n",
+    "]\n",
+    "\n",
+    "# Test with a math problem\n",
+    "calculator_response = run_with_tools(\n",
+    "    \"What is 1337 multiplied by 42?\", \n",
+    "    calculator_tools\n",
+    ")\n",
+    "\n",
+    "print(json.dumps(calculator_response.model_dump(), indent=2))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Tool Response Processing\n",
+    "\n",
+    "Once the model pauses with its `tool_calls` block we, not the LLM, take the wheel.\n",
+    "The helper below does four things in one sweep:\n",
+    "\n",
+    "1. **Parse the call**: It inspects `response.choices[0].message.tool_calls`, pulls out the function name, and JSON-decodes the arguments\n",
```

Comment:
Careful with format, unsure if we put bold text followed by ":", capitalizing the first letter after ":" etc